### ISSUE TITLE
Running R-CNN output mAP=0.52 even if example outputs 0.82

### ISSUE THREAD

**Issue by mortengryning** [Microsoft/CNTK#1154]
Hi.

I'm trying CNTK. I tried the R-CNN example and ran the python scripts. While the network seemed to work, I only got a mAP of 0.52 even if the example (https://github.com/Microsoft/CNTK/wiki/Object-Detection-using-Fast-R-CNN) said 0.8688. When I execute the B3_visualizeoutputroi it only finds tomatos and peppers, even if the example page show all items (milk, mustard, etc.)

![image](https://cloud.githubusercontent.com/assets/7712296/21041657/c581e158-bdec-11e6-9326-e00d66cb7beb.png)

Any idea why?

**Comment by mortengryning** [comment]
When i try with the pre-defined model i get this:

![image](https://cloud.githubusercontent.com/assets/7712296/21050012/84478496-be18-11e6-9c6a-b0e20862ec40.png)

I tried to reinstall everything and do the training again. It seems as I get a good error percentage 

![from_train_grocery](https://cloud.githubusercontent.com/assets/7712296/21050032/b196549a-be18-11e6-98e3-8d30b3b89154.PNG)

But when I evaluate the model using A3_... I still get 0.50 mAP.

**Comment by mortengryning** [comment]
I have now tried both the CPU/GPU version and on two different computers. Is the example wrong?

**Comment by KolyaIvankov** [comment]
I'd like to second the issue, from another point of view.

My example is an adaptation of RegrSimple_CIFAR10.cntk for my own set of images, the task is sto determine the bounding box of a particular object that is always present at every image. The values of the bound vary in the interval from 70 to 150. I have about several hundred training and test images. Here is a snippet of my BrainScript code:

```
command = TrainConvNet:Write:Test

makeMode = false ; traceLevel = 1 ; deviceId = -1

rootDir = ".." ; dataDir  = "$RootDir$/Haystack" ;
outputDir = "Output" ; modelDir = "$outputDir$/Models"

modelPath = "$ModelDir$/RegrSimple_Needle.cmf"

# Training action for a convolutional network
TrainConvNet = {
    action = "train"

    BrainScriptNetworkBuilder = [
        imageShape = 150:200:3
        featScale = Constant(1/256)
        labelDim = 1

        model (features) = {
            featNorm = Scale(features, featScale)
            h1 = LinearLayer {200,        init="gaussian", initValueScale=1.5} (featNorm)
            ol = LinearLayer {labelDim, init="gaussian", initValueScale=1.5} (h1)
        }.ol

        # inputs
        features = Input {imageShape}
        regrLabels = Input {labelDim}
        
        # apply model to features
        ol = model (features)

        # define regression loss
        # rmse = sqrt(SquareError(regrLabels, ol) / labelDim)
        sqerr = SquareError (regrLabels, ol)
        rmse = Sqrt (Constant(1/labelDim).* sqerr)

        featureNodes    = (features)
        labelNodes      = (regrLabels)
        criterionNodes  = (rmse)
        evaluationNodes = (rmse)
        outputNodes     = (ol)
    ]

    SGD = {
        epochSize = 0

        maxEpochs = 200
        minibatchSize = 128
        learningRatesPerSample = 0.0005
        momentumAsTimeConstant = 1024

        firstMBsToShowResult = 5 ; numMBsToShowResult = 50
    }
    reader = {
        verbosity = 0 ; randomize = true
        deserializers = ({
            type = "ImageDeserializer" ; module = "ImageReader"
            file = "$dataDir$/train_map.txt"
            input = {
                features = { transforms = (
                    { type = "Scale" ; width = 150 ; height = 200 ; channels = 3 ; interpolations = "linear" } :
                    { type = "Transpose" }
                )}
                ignored = { labelDim = 10 }
            }
        } : {
            type = "CNTKTextFormatDeserializer" ; module = "CNTKTextFormatReader"
            file = "$dataDir$/train_regrLabels.txt"
            input = {
                regrLabels = { dim = 1 ; format = "dense" }
            }
        })
    }
}
```
Since the database is not too large and the network is quite simple, I am able to make these 200 epochs in a short time, after which I am promised to rmse (which is middle squares) of about 0.7 on the training data and about 1.7 on the test data. But when I then include the model via NuGet package into my C# project, then even on the test data the evaluation gives numbers varying times as much:  I still need to establish the actual value of the middle square error, but empirically it is about 5-10 with outbursts to up to 30. This would have been understandable if I've been feeding new data to an overfitted network, but, as said, was precisely the data on which the overfitting could have been achieved.

I've been dealing with this problem for several days already, thinking first that I am encoding my pictures by a method different from the one used in the ImageReader module. Since then, I've used both ExtractCHW and ExtractHWC functions from EvalClients project as well as ExtractCWH and ExtractWHC, just to be sure. ExtractCHW seems to be the right one - using the other 3 give just random numbers.

Is something wrong with statistics, which is possibly left unnoticed when people do classification? Does SquareError work correctly, or (more likely) am I misunderstanding something?

**Comment by mortengryning** [comment]
KolyaIvankov I think it would be more beneficial if you created your own post.

I "solved" this issue. I used the 2.0 beta 3 (instead of 5) and got an accuracy of 0.889. It seem as there is an issue in the never beta 4+5 that you should correct ASAP so other do not get the same error. I will leave this issue open.

**Comment by cha-zhang** [comment]
I tested with beta 6.0 just now and I confirmed again that I got 0.8688 on my Windows machine. This mAP number could fluctuate if the network contains some random initialization or sample permutation. So your number could be different from mine.

**Comment by mogrysama** [comment]
I have the same problem like you.
I always get a mAP of 0.5032. After setting image_set = 'train' I get 0.4875. I´m using CNTK for Windows v.2.0 Beta 6 CPU only and did everything like it is wrtitten here. Can anybody help?
And if I use the pretrained model I get 0.0000 :-/

**Comment by mortengryning** [comment]
try to use beta 3.0 instead.

**Comment by mogrysama** [comment]
yes, already tried it with beta3 and it worked fine. like you said. but why it is not working with 4,5 and 6? 6 was released 2 days ago and has the same problem wit the mAP...

**Comment by cha-zhang** [comment]
Are the low numbers all because of CPU? or you get low numbers on GPU as well?

**Comment by mogrysama** [comment]
Only used the CPU. I don´t have a graphical card (nor nvidia) in my workstation. But with beta3 i got the same results, so for now it is fine for me.

**Comment by cha-zhang** [comment]
We have identified the issue caused by a bug in the CPU convolution path when the stride is large and padding is true (which is the case for the particular AlexNet we have now). We will have a fix in the next day or two. Thanks!

**Comment by Vedaevolution** [comment]
I just want to inform you that the bug is still existent in version 2.0 Beta 7. Thank you very much for solving the issue!

Best Regards
 Stephen

**Comment by cha-zhang** [comment]
It's not in beta 7. It's in the current master branch and will be in beta 8.

_____________________________
